eventdisplay/FairEventManagerEditor.cxx

```diff
@@ -111,77 +124,127 @@ void FairEventManagerEditor::Init()
-    fMinEnergy->SetNELength(5);
-    fMinEnergy->SetLabelWidth(80);
-    fMinEnergy->Build();
-    fMinEnergy->SetLimits(0, MAXE, 2501, TGNumberFormat::kNESRealOne);
-    fMinEnergy->SetToolTip("Minimum energy of displayed track.");
-    fMinEnergy->Connect("ValueSet(Double_t)", "FairEventManagerEditor", this, "MinEnergy()");
-    title1->AddFrame(fMinEnergy, new TGLayoutHints(kLHintsTop, 1, 1, 1, 0));
-
-    fMaxEnergy = new TEveGValuator(title1, "Max Energy:", 90, 0);
-    fMaxEnergy->SetNELength(5);
-    fMaxEnergy->SetLabelWidth(80);
-    fMaxEnergy->Build();
-    fMaxEnergy->SetLimits(0, MAXE, 2501, TGNumberFormat::kNESRealOne);
-    fMaxEnergy->SetToolTip("Maximum energy of displayed track.");
-    fMaxEnergy->SetValue(MAXE);
-    fMaxEnergy->Connect("ValueSet(Double_t)", "FairEventManagerEditor", this, "MaxEnergy()");
-    title1->AddFrame(fMaxEnergy, new TGLayoutHints(kLHintsTop, 1, 1, 1, 0));
-
     TGTextButton* fUpdate = new TGTextButton(title1, "Update");
     title1->AddFrame(fUpdate, new TGLayoutHints(kLHintsRight | kLHintsExpandX, 5, 5, 1, 1));
     fUpdate->Connect("Clicked()", "FairEventManagerEditor", this, "SelectEvent()");
 
     fInfoFrame->AddFrame(title1, new TGLayoutHints(kLHintsTop, 0, 0, 2, 0));
 
     fManager->SetMaxEnergy(MAXE);
-}
 
-void FairEventManagerEditor::MaxEnergy() { fManager->SetMaxEnergy(fMaxEnergy->GetValue()); }
-
-void FairEventManagerEditor::MinEnergy() { fManager->SetMinEnergy(fMinEnergy->GetValue()); }
-
-void FairEventManagerEditor::DoVizPri()
-{
-    if (fVizPri->IsOn()) {
-        fManager->SetPriOnly(kTRUE);
-    } else {
-        fManager->SetPriOnly(kFALSE);
-    }
+    //=============== graphics =============================
+    TGVerticalFrame* scene_conf = CreateEditorTabSubFrame("Graphics");
+    TGHorizontalFrame* transparency_frame = new TGHorizontalFrame(scene_conf);
+    TGCheckButton* Transparency = new TGCheckButton(transparency_frame, "Global transparency");
+    transparency_frame->AddFrame(Transparency, new TGLayoutHints(kLHintsLeft | kLHintsExpandX, 5, 5, 1, 1));
+    Transparency->Connect("Toggled(Bool_t)", this->ClassName(), this, "SwitchTransparency(Bool_t)");
+
+    fGlobalTransparency = new TGNumberEntry(transparency_frame,
+                                            0.,
+                                            6,
+                                            -1,
+                                            TGNumberFormat::kNESInteger,
+                                            TGNumberFormat::kNEANonNegative,
+                                            TGNumberFormat::kNELLimitMinMax,
+                                            0,
+                                            100);
+    fGlobalTransparency->SetIntNumber(80);
+    transparency_frame->AddFrame(fGlobalTransparency, new TGLayoutHints(kLHintsRight, 1, 1, 1, 1));
+    fGlobalTransparency->Connect("ValueSet(Long_t)", "FairEventManagerEditor", this, "SwitchTransparency(Bool_t)");
```

Comment:
Something with this observer goes wrong. The `Long_t` argument seems to not contain the transparency number, when clicking the increase button.

For a fix I propose something like the following: https://github.com/dennisklein/FairRoot/commit/3f13e209257facf4f678f86eed70d71bd28b11aa

Reply:
This ValueSet is only to be sure that function will be called when user change the value (fCurrentEvent works in similar way). According to documentation it contains information about step (or something like that, but not about the stored value).
Also in ROOT tutorials (https://root.cern/doc/master/numberEntry_8C.html ) GetValue(Long_t) is not used to get value from button. I don't think that creation of new class only for a single button (and basically no extra functionality) is a good idea.
Btw. what is the point of using unique_pointer and call release method few lines after?

Reply:
> I don't think that creation of new class only for a single button (and basically no extra functionality) is a good idea.

Why?

> (and basically no extra functionality)

My example has quite some of different functionality. It creates a new signal that consolidates the two signal from the check box and from the number entry. This signal only fires if there is a meaningful change and only presents a single int parameter (instead of bool + int). This signal is also wired up directly with the manager and not in-directed throw another function in the editor.

> Btw. what is the point of using unique_pointer and call release method few lines after?

* Exception-safety: If your programs throws an exception in these few lines, the object will not leak. (stack unwinds and dtor of the unique ptr will delete the object)
* Readability: The code shows clearly and unambiguously who owns the objects.

---

Anyways, if I check your box and then click the arrow up button on the number entry widget, it does not behave as expected. It behaves as if I un-check the box.

Reply:
Now I think it's fixed - however I slightly modified your class. I also moved all button-like classes to eventdisplay/gui directory (I suppose we will get more of them in the future).
Thank you for noticing this bug.